packages/dataviews/src/view-grid.js

```diff
@@ -7,20 +7,223 @@ import classnames from 'classnames';
 import ItemActions from './item-actions';
 import SingleSelectionCheckbox from './single-selection-checkbox';
 
+const {
+	useCompositeStoreV2: useCompositeStore,
+	CompositeV2: Composite,
+	CompositeItemV2: CompositeItem,
+	CompositeRowV2: CompositeRow,
+} = unlock( componentsPrivateApis );
+
+const GridContext = createContext( {} );
+
+/**
+ * Split flat array of cells into groups.
+ *
+ * @param {Array}   cells     Flat array of cells to split
+ * @param {integer} chunkSize Size to split cells
+ * @return {Array} Cells split by `chunkSize`
+ */
+function useChunkedCells( cells, chunkSize ) {
+	return useMemo( () => {
+		const chunks = [];
+
+		if ( cells?.length && chunkSize ) {
+			for ( let i = 0; i < cells.length; i += chunkSize ) {
+				chunks.push( cells.slice( i, i + chunkSize ) );
+			}
+		}
+
+		return chunks;
+	}, [ cells, chunkSize ] );
+}
+
+/**
+ * Hook to automatically convert flat list of child cells into rows,
+ * determined by the number of columns that fit into the available space.
+ *
+ * @param {Array}      cells     Flat array of cells to use in rows
+ * @param {ElementRef} targetRef Reference to determine column width
+ * @return {JSXElement} Rendered rows
+ */
+function useRows( cells, targetRef ) {
+	const [ columnCount, setColumnCount ] = useState( cells?.length || 1 );
+	const [ resizeListener, { width: totalWidth } ] = useResizeObserver();
+	const chunkedCells = useChunkedCells( cells, columnCount );
+
+	const referenceCell =
+		targetRef?.current?.querySelector( '[role="gridcell"]' ) || {};
+	const { offsetWidth: columnWidth = totalWidth } = referenceCell;
+
+	useEffect( () => {
+		if ( columnWidth && totalWidth ) {
+			setColumnCount(
+				Math.max( 1, Math.floor( totalWidth / columnWidth ) )
+			);
+		}
+	}, [ columnWidth, totalWidth ] );
+
+	const rows = useMemo(
+		() =>
+			chunkedCells.map( ( row, index ) => (
+				<CompositeRow
+					className="dataviews-view-grid__row"
+					role="row"
+					key={ `row-${ index }` }
+				>
+					{ row }
+				</CompositeRow>
+			) ),
+
+		[ chunkedCells ]
+	);
+
+	return useMemo(
+		() => (
+			<>
+				{ rows }
+				{ resizeListener }
+			</>
+		),
+		[ resizeListener, rows ]
+	);
+}
+
+/**
+ * Create CSS selector for element. We avoid IDs and classes as
+ * they can be unstable across renders.
+ *
+ * @param {string}         context Grid item identifier
+ * @param {Element | null} element Node to get selector for
+ * @return {string | undefined} Selector for element
+ */
+function getFocusSelector( context, element ) {
+	if ( ! element || ! element.parentElement ) return undefined;
+	if ( element.id && element.id === context ) {
+		return `#${ context.replaceAll( '/', '\\/' ) }`;
+	}
+
+	const parent = getFocusSelector( context, element.parentElement );
+	const siblings = Array.from( element.parentElement.children );
+	const index = siblings.indexOf( element ) + 1;
+
+	return `${ parent } > ${ element.nodeName }:nth-child(${ index })`;
```

Comment:
I'm still reviewing and there are lots of things going on here, so sorry if some of my questions are obvious or wrong, but I want to understand fully what's going on :). 

Why do we need create a selector and then query the `doc` to focus(which is expensive)? Couldn't we do that by keeping track of the ref?